build failed on windows 10

```r
sessionInfo()
```
```
R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)
```

Then try to install using devtools
```r
devtools::install_github("rstudio/websocket")
```
```
Downloading GitHub repo rstudio/websocket@master
from URL https://api.github.com/repos/rstudio/websocket/zipball/master
WARNING: Rtools 3.5 found on the path at C:/Programs/Rtools is not compatible with R 3.5.0.

Please download and install the appropriate version of Rtools from http://cran.r-project.org/bin/windows/Rtools/, remove the incompatible version from your PATH.
Installing websocket
"c:/Programs/R/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/dmccauley/AppData/Local/Temp/RtmpMrkbq6/devtools6242eb347f0/rstudio-websocket-9bf435f"  \
  --library="c:/Programs/R/library" --install-tests 

* installing *source* package 'websocket' ...

   **********************************************
   WARNING: this package has a configure script
         It probably needs manual configuration
   **********************************************


** libs

*** arch - i386
rm -f websocket.dll RcppExports.o websocket.o
"c:/Programs/R/bin/i386/Rscript.exe" "../tools/winlibs.R" 1.0.2o
C:/Programs/Rtools/mingw_32/bin/g++  -std=gnu++11 -I"c:/Programs/R/include" -DNDEBUG -I./lib -I../windows/openssl-1.0.2o/include -D_WEBSOCKETPP_CPP11_THREAD_ -I"c:/Programs/R/library/Rcpp/include" -I"c:/Programs/R/library/BH/include" -I"c:/Programs/R/library/AsioHeaders/include"        -O2 -Wall  -mtune=generic -c RcppExports.cpp -o RcppExports.o
C:/Programs/Rtools/mingw_32/bin/g++  -std=gnu++11 -I"c:/Programs/R/include" -DNDEBUG -I./lib -I../windows/openssl-1.0.2o/include -D_WEBSOCKETPP_CPP11_THREAD_ -I"c:/Programs/R/library/Rcpp/include" -I"c:/Programs/R/library/BH/include" -I"c:/Programs/R/library/AsioHeaders/include"        -O2 -Wall  -mtune=generic -c websocket.cpp -o websocket.o
In file included from ./lib/websocketpp/config/asio_no_tls_client.hpp:32:0,
                 from client.hpp:4,
                 from websocket.cpp:24:
./lib/websocketpp/transport/asio/endpoint.hpp: In member function 'void ws_websocketpp::transport::asio::endpoint<config>::handle_accept(ws_websocketpp::transport::accept_handler, const error_code&)':
./lib/websocketpp/transport/asio/endpoint.hpp:839:28: error: 'operation_canceled' is not a member of 'std::errc'
             if (asio_ec == lib::asio::errc::operation_canceled) {
                            ^
make: *** [c:/Programs/R/etc/i386/Makeconf:215: websocket.o] Error 1
ERROR: compilation failed for package 'websocket'
* removing 'c:/Programs/R/library/websocket'
In R CMD INSTALL
Installation failed: Command failed (1)
```

I have seen the reference to asio headers  in https://github.com/rstudio/websocket/issues/15#issuecomment-418506229
 and installed these without success - am I missing an install for errc?

Thanks
Damien

--- a/src/lib/update.sh
+++ b/src/lib/update.sh
@@ -18,3 +18,6 @@ find . -path ./.git -prune -o -type f -print0 | xargs -0 sed -i "" -e 's/&std::c
 cd ../..
 mv websocketpp_repo/websocketpp .
 rm -rf websocketpp_repo
+
+echo "IMPORTANT NOTE: Apply this patch manually:" >&2
+echo "https://github.com/rstudio/websocket/commit/063ca452c7639b952dfd4981602436d43305457a" >&2

--- a/src/lib/websocketpp/transport/asio/endpoint.hpp
+++ b/src/lib/websocketpp/transport/asio/endpoint.hpp
@@ -45,6 +45,69 @@ namespace ws_websocketpp {
 namespace transport {
 namespace asio {
 
+/***
+ * In current versions of libstdc++, the error_constants.h code associated to mingw32
+ * does not define certain standard enum values for `std::errc`. (In C++11 standard,
+ * sections 19.5.2, 19.5.3.) Asio uses these for lib::asio::errc when it is compiled
+ * as a stand-alone library, so because of the libstdc++ defect, code below referring
+ * to lib::asio::errc::operation_canceled fails to compile on mingw.
+ *
+ * This workaround detects the defect using SFINAE and returns 'false' for the check
+ * if operation_canceled is not defined, instead of failing to compile.
+ *
+ * If upstream patches this later by defining those enum values, then the workaround
+ * will stop having any effect.
+ *
+ * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68307
+ */
+namespace _workaround_gcc_libstdcpp_issue_68307_missing_values {
+/***
+ * Same as std::enable_if, but don't want to introduce dependency on <type_traits>
+ * since that's C++11 only
+ */
+template<bool B, class T = void>
+struct enable_if {};
+
+template<class T>
+struct enable_if<true, T> { typedef T type; };
+
+/***
+ * Metafunction to test "operation_canceled" value
+ */
+template <typename T, typename ENABLE=void>
+struct op_canceled_helper {
+  template <typename U>
+  static inline bool is_op_canceled(const U & u) { return false; }
+};
+
+template <typename T>
+struct op_canceled_helper<T, enable_if<T::operation_canceled == T::operation_canceled, void> > {
+  template<typename U>
+  static inline bool is_op_canceled(const U & u) { return u == T::operation_canceled; }
+};
+
+/***
+ * This function is intended to be a drop-in replacement for
+ *   (asio_ec == lib::asio::errc::operation_canceled)
+ *
+ * except that if lib::asio::errc::operation_canceled does not exist, it returns false,
+ * instead of failing to compile.
+ *
+ * When using boost and not asio standalone, then lib::asio::errc is a namespace, not an enum class.
+ * So the template code will fail to compile and we need to block it from being instantiated, with this
+ * ifdef. When using boost the standard library symbol definitions aren't relevant afaik.
+ */
+#ifdef ASIO_STANDALONE
+static inline bool is_op_canceled(const lib::asio::error_code & asio_ec) {
+  return op_canceled_helper<lib::asio::errc, void>::is_op_canceled(asio_ec);
+}
+#else
+static inline bool is_op_canceled(const lib::asio::error_code & asio_ec) {
+  return asio_ec == lib::asio::errc::operation_canceled;
+}
+#endif
+} // namespace _workaround
+
 /// Asio based endpoint transport component
 /**
  * transport::asio::endpoint implements an endpoint transport component using
@@ -161,7 +224,7 @@ public:
             rhs.m_acceptor = NULL;
             rhs.m_listen_backlog = lib::asio::socket_base::max_connections;
             rhs.m_state = UNINITIALIZED;
-            
+
             // TODO: this needs to be updated
         }
         return *this;
@@ -226,7 +289,7 @@ public:
      * @param ec Set to indicate what error occurred, if any.
      */
     void init_asio(lib::error_code & ec) {
-        // Use a smart pointer until the call is successful and ownership has 
+        // Use a smart pointer until the call is successful and ownership has
         // successfully been taken. Use unique_ptr when available.
         // TODO: remove the use of auto_ptr when C++98/03 support is no longer
         //       necessary.
@@ -248,7 +311,7 @@ public:
      * @see init_asio(io_service_ptr ptr)
      */
     void init_asio() {
-        // Use a smart pointer until the call is successful and ownership has 
+        // Use a smart pointer until the call is successful and ownership has
         // successfully been taken. Use unique_ptr when available.
         // TODO: remove the use of auto_ptr when C++98/03 support is no longer
         //       necessary.
@@ -379,7 +442,7 @@ public:
     lib::asio::io_service & get_io_service() {
         return *m_io_service;
     }
-    
+
     /// Get local TCP endpoint
     /**
      * Extracts the local endpoint from the acceptor. This represents the
@@ -387,7 +450,7 @@ public:
      *
      * Sets a bad_descriptor error if the acceptor is not currently listening
      * or otherwise unavailable.
-     * 
+     *
      * @since 0.7.0
      *
      * @param ec Set to indicate what error occurred, if any.
@@ -426,10 +489,10 @@ public:
 
         m_acceptor->open(ep.protocol(),bec);
         if (bec) {ec = clean_up_listen_after_error(bec);return;}
-        
+
         m_acceptor->set_option(lib::asio::socket_base::reuse_address(m_reuse_addr),bec);
         if (bec) {ec = clean_up_listen_after_error(bec);return;}
-        
+
         // if a TCP pre-bind handler is present, run it
         if (m_tcp_pre_bind_handler) {
             ec = m_tcp_pre_bind_handler(m_acceptor);
@@ -438,13 +501,13 @@ public:
                 return;
             }
         }
-        
+
         m_acceptor->bind(ep,bec);
         if (bec) {ec = clean_up_listen_after_error(bec);return;}
-        
+
         m_acceptor->listen(m_listen_backlog,bec);
         if (bec) {ec = clean_up_listen_after_error(bec);return;}
-        
+
         // Success
         m_state = LISTENING;
         ec = lib::error_code();
@@ -828,7 +891,7 @@ protected:
         m_elog = e;
     }
 
-    void handle_accept(accept_handler callback, lib::asio::error_code const & 
+    void handle_accept(accept_handler callback, lib::asio::error_code const &
         asio_ec)
     {
         lib::error_code ret_ec;
@@ -836,7 +899,7 @@ protected:
         m_alog->write(log::alevel::devel, "asio::handle_accept");
 
         if (asio_ec) {
-            if (asio_ec == lib::asio::errc::operation_canceled) {
+            if (_workaround_gcc_libstdcpp_issue_68307_missing_values::is_op_canceled(asio_ec)) {
                 ret_ec = make_error_code(ws_websocketpp::error::operation_canceled);
             } else {
                 log_err(log::elevel::info,"asio handle_accept",asio_ec);

